plugins/amqp/src/plugin.cpp

```diff
@@ -600,8 +624,21 @@ class rabbitmq_saver final : public plugin_saver {
       .mandatory = args_.mandatory,
       .immediate = args_.immediate,
     };
-    return [&ctrl, engine = std::make_shared<amqp_engine>(std::move(*engine)),
-            opts = std::move(opts)](chunk_ptr chunk) mutable {
+    auto hb = try_get<uint64_t>(config_, "heartbeat");
+    if (hb && *hb && **hb > 0) {
+      // If we are requesting heartbeats, we are also responsible to handle the
+      // heartbeats we get. If we have long gaps in interaction with the broker,
+      // we need to proactively check explicitly if there is something for us.
+      // We check 3 times per heartbeat interval, at most once per second.
+      auto interval = std::max(uint64_t{1}, **hb / 3);
+      TENZIR_DEBUG("using heartbeat interval of {} seconds", interval);
+      detail::weak_run_delayed_loop(&ctrl.self(),
```

Comment:
Is this thread-safe? (I haven't used this CAF construct, so I dunno.) But we're accessing shared state of the engine and calling into a member function that accesses `conn_`.

Reply:
The `weak_run_delayed_loop` lambda runs in the context of the actor.